<a href="https://colab.research.google.com/github/YaroslavTrusov/python-ai-template-Yaroslav_Trusov/blob/main/notebooks/week2b_read_csv.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 📊 Week 2: Data Analysis — Чтение и проверка данных

**Цель**: Научиться читать CSV-файлы из репозитория GitHub в Google Colab и выполнять базовую проверку данных с помощью pandas.

**Данные:**
- [`planes.csv`](https://github.com/YaroslavTrusov/python-ai-template-Yaroslav_Trusov/blob/main/data/planes.sparql) — виды самолетов, двигателей, цена, максимальная высота и суммарное количество, произведенных самолетов
- [`engines.csv`](https://github.com/YaroslavTrusov/python-ai-template-Yaroslav_Trusov/blob/main/data/engines.sparql) — виды двигателей, используемых в самолётах, с обязательным производителем и страной
- [`engine_manufacturers.csv`](https://github.com/YaroslavTrusov/python-ai-template-Yaroslav_Trusov/blob/main/data/engine_manufacturers.sparql) — производители двигателей, используемых в самолётах, с метками страны и штаб-квартиры


**Что мы делаем:**
1. Клонируем репозиторий GitHub в Colab
2. Читаем CSV-файлы в pandas DataFrame
3. Очищаем и переименовываем столбцы
4. Смотрим структуру данных и делаем быструю валидацию

## 🐱 [1] Клонируем репозиторий курса в Colab

In [33]:
# 🐱 Шаг 1. Клонируем репозиторий в Colab

import os

repo = "python-ai-template-Yaroslav_Trusov"
repo_path = f"/content/{repo}"  # абсолютный путь — не зависит от cwd

if not os.path.exists(repo_path):          # всегда проверяет /content/python-ai-template
    !git clone -q https://github.com/YaroslavTrusov/python-ai-template-Yaroslav_Trusov

if os.getcwd() != repo_path:               # точное сравнение, не endswith
    %cd {repo_path}

print("✅ Репозиторий готов, теперь мы работаем внутри папки", repo)

✅ Репозиторий готов, теперь мы работаем внутри папки python-ai-template-Yaroslav_Trusov


## 📥 [2A] Простое чтение CSV-файлов в pandas

Сначала просто прочитаем CSV-файл в объект `DataFrame`, без каких‑либо изменений.

После этого мы узнаем, сколько строк загружено в датасет.

In [34]:
# 🐱 Шаг 2A. Чтение CSV-файла в pandas

import pandas as pd
from IPython.display import display  # Импортируем функцию display

df_planes = pd.read_csv("data/planes.csv")
df_engines = pd.read_csv("data/engines.csv")
df_engine_manufacturers = pd.read_csv("data/engine_manufacturers.csv")

print("✅ Загружено строк в df_planes:", len(df_planes))
print("✅ Загружено строк в df_engines:", len(df_engines))
print("✅ Загружено строк в df_engine_manufacturers:", len(df_engine_manufacturers))

# Отображаем первые 5 строк каждого датафрейма
display(df_planes.head())
display(df_engines.head())
display(df_engine_manufacturers.head())




✅ Загружено строк в df_planes: 1109
✅ Загружено строк в df_engines: 1224
✅ Загружено строк в df_engine_manufacturers: 309


,aircraft,aircraftLabel,engine,engineLabel,maxAltitude,price,totalProduced,firstFlight,countryLabel,manufacturerLabel
0,http://www.wikidata.org/entity/Q211443,Ил-2,http://www.wikidata.org/entity/Q599854,АМ-38,NaN,NaN,36163,1939-10-02T00:00:00Z,NaN,Авиационный комплекс имени С. В. Ильюшина
1,http://www.wikidata.org/entity/Q943360,У-2,http://www.wikidata.org/entity/Q742981,М-11,NaN,NaN,33528,1927-06-24T00:00:00Z,NaN,ЦКБ-39
2,http://www.wikidata.org/entity/Q207700,Антонов Ан-2,http://www.wikidata.org/entity/Q15051031,Shvetsov ASh-62IR,NaN,NaN,18000,1947-08-31T00:00:00Z,СССР,Hongdu Aviation Industry Group
3,http://www.wikidata.org/entity/Q562475,Як-9,http://www.wikidata.org/entity/Q977265,М-105,10750.0,NaN,16769,1942-01-01T00:00:00Z,NaN,Омск
4,http://www.wikidata.org/entity/Q562475,Як-9,http://www.wikidata.org/entity/Q977265,М-105,10750.0,NaN,16769,1942-10-01T00:00:00Z,NaN,Омск


,engine,engineLabel,manufacturerLabel,countryLabel,basedOnLabel,cylinders,fuelOrCoolantLabel
0,http://www.wikidata.org/entity/Q106620985,Q106620985,Horch,Германия,NaN,8.0,жидкая вода
1,http://www.wikidata.org/entity/Q131357998,Q131357998,Volvo Penta,Швеция,NaN,6.0,NaN
2,http://www.wikidata.org/entity/Q133886954,Q133886954,Volvo Penta,Швеция,NaN,6.0,NaN
3,http://www.wikidata.org/entity/Q134479603,Q134479603,Electro-Motive Diesel,США,NaN,16.0,NaN
4,http://www.wikidata.org/entity/Q25399814,Q25399814,Avia,Чехия,NaN,9.0,NaN


,manufacturer,manufacturerLabel,countryLabel,headquartersLabel,coordinates
0,http://www.wikidata.org/entity/Q2308012,Q2308012,Франция,Мольсем,NaN
1,http://www.wikidata.org/entity/Q2308012,Q2308012,Франция,Дорлисайм,NaN
2,http://www.wikidata.org/entity/Q26678,Q26678,Германия,Мюнхен,NaN
3,http://www.wikidata.org/entity/Q9380476,Q9380476,Польша,Калиш,NaN
4,http://www.wikidata.org/entity/Q262737,ABC Motors,Великобритания,Хершам,NaN


## 🧹 [2B] Очистка и переименование столбцов

В исходных CSV-файлах есть **технические столбцы**, которые полезны для Викиданных, но мешают простому анализу:

- Столбец `aircraft` с URL (ссылкой на объект Wikidata) — **сохраняем его для отладки**, но переименуем в `aircraftURL`.
- Столбцы `aircraftLabel`, `engine`, `engineLabel`, `maxAltitude`, `price`, `totalProduced` содержат читаемые подписи (названия).

В этом шаге мы:
- переименуем столбец с URL Wikidata (`aircraft` → `aircraftURL`);
- переименуем `aircraftLabel → aircraft`, `engine → engineURL`, `engineLabel → engine`, `maxAltitude → altitude`, `totalProduced → produced`;
- приведём числовые столбцы (`altitude`,`price` `produced`) к типу `int`.

При приведении к числам мы используем:

- `pd.to_numeric(..., errors="coerce")` — преобразует значения в числа, некорректные значения превращает в `NaN`;
- `fillna(0)` — заменяет пропущенные значения (`NaN`) на 0;
- `astype(int)` — переводит столбец к целочисленному типу.

> ⚠️ **Важно:** если в ваших данных есть столбцы с URL Wikidata и столбцы вида `*Label`, этот шаг обязателен, чтобы получить аккуратные таблички для анализа. Столбец `URL` пригодится, если нужно будет быстро перейти к оригинальной записи в Викиданных.

In [35]:
# 🧹 ШАГ 2B. Очистка и переименование столбцов

# ============================================================================
# 1️⃣ df_planes: Самолёты
# ============================================================================
print("=" * 60)
print("✈️ Обработка df_planes (самолёты)")
print("=" * 60)

# 🔍 Диагностика: какие столбцы есть в df_planes?
print("\n📋 Доступные столбцы в df_planes:")
print(f"   {df_planes.columns.tolist()}")

# 🔧 Поиск столбца двигателя среди всех возможных вариантов
possible_engine_label_cols = ['engineLabel', 'engine_label', 'EngineLabel', 'engine_nameLabel']
engine_label_found = None
for col in possible_engine_label_cols:
    if col in df_planes.columns:
        engine_label_found = col
        print(f"   ✅ Найден столбец двигателя: {col}")
        break

if engine_label_found is None:
    print("   ⚠️ Столбец engineLabel не найден! Проверьте исходный CSV")

# 🔧 Поиск столбца engine URL
possible_engine_url_cols = ['engine', 'engineURL', 'engine_url', 'Engine']
engine_url_found = None
for col in possible_engine_url_cols:
    if col in df_planes.columns and col != engine_label_found:
        engine_url_found = col
        print(f"   ✅ Найден столбец engine URL: {col}")
        break

# Переименование столбцов (с проверкой наличия)
rename_dict = {}

# Самолёт: URL и название
if "aircraftLabel" in df_planes.columns:
    rename_dict["aircraftLabel"] = "aircraft"
if "aircraft" in df_planes.columns and "aircraftLabel" in df_planes.columns:
    rename_dict["aircraft"] = "aircraftURL"
elif "URL" in df_planes.columns and "aircraftURL" not in df_planes.columns:
    rename_dict["URL"] = "aircraftURL"

# Двигатель — явное переименование engineLabel → engine
if engine_label_found:
    rename_dict[engine_label_found] = "engine"
    print(f"   → Будет переименован: {engine_label_found} → engine")

if engine_url_found and engine_url_found != engine_label_found:
    rename_dict[engine_url_found] = "engineURL"
    print(f"   → Будет переименован: {engine_url_found} → engineURL")

# Остальные столбцы
if "maxAltitude" in df_planes.columns:
    rename_dict["maxAltitude"] = "altitude"
if "totalProduced" in df_planes.columns:
    rename_dict["totalProduced"] = "produced"

# Выполнение переименования
if rename_dict:
    df_planes = df_planes.rename(columns=rename_dict)
    print(f"\n✅ Переименовано столбцов: {len(rename_dict)}")
    for old_name, new_name in rename_dict.items():
        print(f"   • {old_name:20s} → {new_name}")
else:
    print("\n⏭️ Столбцы уже переименованы, пропускаем")

# 🔍 ПРОВЕРКА: появился ли столбец 'engine'?
print(f"\n🔍 ПРОВЕРКА РЕЗУЛЬТАТА:")
if "engine" in df_planes.columns:
    print(f"   ✅ СТОЛБЕЦ 'engine' СОЗДАН!")
    print(f"   → Тип: {df_planes['engine'].dtype}")
    print(f"   → Заполнено: {df_planes['engine'].notna().sum()} из {len(df_planes)}")
else:
    print(f"   ❌ СТОЛБЕЦ 'engine' ОТСУТСТВУЕТ!")
    print(f"   → Доступные столбцы: {df_planes.columns.tolist()}")

# Проверка заполненности ДО очистки (OPTIONAL-поля)
print("\n📊 Заполненность данных ДО очистки:")
optional_planes = ["price", "altitude", "produced"]
for col in optional_planes:
    if col in df_planes.columns:
        fill_rate = df_planes[col].notna().mean()
        status = "⚠️ НИЗКАЯ" if fill_rate < 0.15 else "✅"
        print(f"   {status} {col:12s}: {fill_rate:6.2%} ({df_planes[col].notna().sum()} из {len(df_planes)})")

# 🔧 Очистка числовых полей
print("\n🔧 Очистка числовых полей:")
for col in ["altitude", "price", "produced"]:
    if col in df_planes.columns:
        if df_planes[col].dtype not in ['Int64', 'int64']:
            df_planes[col] = pd.to_numeric(df_planes[col], errors="coerce").convert_dtypes()
            print(f"   ✅ {col:12s}: преобразован в {df_planes[col].dtype} (NaN сохранён)")
        else:
            print(f"   ⏭️ {col:12s}: уже числовой, пропускаем")

# Проверка заполненности ПОСЛЕ очистки
print("\n📊 Заполненность данных ПОСЛЕ очистки:")
for col in optional_planes:
    if col in df_planes.columns:
        fill_rate = df_planes[col].notna().mean()
        print(f"   {col:12s}: {fill_rate:6.2%}")

# Документирование «длинного формата»
print(f"\n📏 Структура данных:")
print(f"   Уникальных самолётов (aircraftURL): {df_planes['aircraftURL'].nunique()}")
print(f"   Всего строк: {len(df_planes)}")
print(f"   → Среднее строк на самолёт: {len(df_planes)/df_planes['aircraftURL'].nunique():.2f}")
if len(df_planes) > df_planes['aircraftURL'].nunique() * 1.5:
    print("   ⚠️ Данные в «длинном формате» — один самолёт имеет несколько записей")


# ============================================================================
# 2️⃣ df_engines: Двигатели
# ============================================================================
print("\n" + "=" * 60)
print("🔧 Обработка df_engines (двигатели)")
print("=" * 60)

# 🔍 ИСПРАВЛЕНИЕ: Поиск столбца fuelOrCoolantLabel среди возможных вариантов
possible_fuel_cols = ['fuelOrCoolantLabel', 'fuelOrCoolant', 'cooling', 'coolingType', 'fuelType']
fuel_col_found = None
for col in possible_fuel_cols:
    if col in df_engines.columns:
        fuel_col_found = col
        print(f"   ✅ Найден столбец топлива/охлаждения: {col}")
        break

if fuel_col_found is None:
    print("   ⚠️ Столбец fuelOrCoolantLabel не найден")

# Переименование столбцов
rename_dict_engines = {}

if "engineLabel" in df_engines.columns:
    rename_dict_engines["engineLabel"] = "engine"
if "engine" in df_engines.columns and "engineLabel" in df_engines.columns:
    rename_dict_engines["engine"] = "engineURL"
if "manufacturerLabel" in df_engines.columns:
    rename_dict_engines["manufacturerLabel"] = "manufacturer"
if "manufacturer" in df_engines.columns and "manufacturerLabel" in df_engines.columns:
    rename_dict_engines["manufacturer"] = "manufacturerURL"

# 🔧 ИСПРАВЛЕНИЕ: Переименование fuelOrCoolantLabel → fuelOrCoolant
if fuel_col_found:
    rename_dict_engines[fuel_col_found] = "fuelOrCoolant"
    print(f"   → Будет переименован: {fuel_col_found} → fuelOrCoolant")

if rename_dict_engines:
    df_engines = df_engines.rename(columns=rename_dict_engines)
    print(f"\n✅ Переименовано столбцов: {len(rename_dict_engines)}")
    for old_name, new_name in rename_dict_engines.items():
        print(f"   • {old_name:20s} → {new_name}")
else:
    print("\n⏭️ Столбцы уже переименованы, пропускаем")

# Проверка заполненности ДО очистки
print("\n📊 Заполненность данных ДО очистки:")
# 🔧 ИСПРАВЛЕНИЕ: cooling заменён на fuelOrCoolant
optional_engines = ["cylinders", "fuelOrCoolant", "power"]
for col in optional_engines:
    if col in df_engines.columns:
        fill_rate = df_engines[col].notna().mean()
        status = "⚠️ НИЗКАЯ" if fill_rate < 0.15 else "✅"
        print(f"   {status} {col:15s}: {fill_rate:6.2%} ({df_engines[col].notna().sum()} из {len(df_engines)})")

# 🔧 Очистка числовых полей
print("\n🔧 Очистка числовых полей:")
for col in ["cylinders", "power"]:
    if col in df_engines.columns:
        if df_engines[col].dtype not in ['Int64', 'int64']:
            df_engines[col] = pd.to_numeric(df_engines[col], errors="coerce").convert_dtypes()
            print(f"   ✅ {col:15s}: преобразован в {df_engines[col].dtype} (NaN сохранён)")
        else:
            print(f"   ⏭️ {col:15s}: уже числовой, пропускаем")

# 🔧 ИСПРАВЛЕНИЕ: Строковые поля (fuelOrCoolant) — привести к нижнему регистру
if "fuelOrCoolant" in df_engines.columns:
    df_engines["fuelOrCoolant"] = df_engines["fuelOrCoolant"].str.lower().str.strip()
    print("   ✅ fuelOrCoolant: приведён к нижнему регистру")
elif "cooling" in df_engines.columns:  # Резервный вариант
    df_engines["cooling"] = df_engines["cooling"].str.lower().str.strip()
    print("   ✅ cooling: приведён к нижнему регистру (резервное имя)")

# Проверка заполненности ПОСЛЕ очистки
print("\n📊 Заполненность данных ПОСЛЕ очистки:")
for col in optional_engines:
    if col in df_engines.columns:
        fill_rate = df_engines[col].notna().mean()
        print(f"   {col:15s}: {fill_rate:6.2%}")

# Документирование структуры
print(f"\n📏 Структура данных:")
print(f"   Уникальных двигателей (engineURL): {df_engines['engineURL'].nunique()}")
print(f"   Всего строк: {len(df_engines)}")
print(f"   → Среднее строк на двигатель: {len(df_engines)/df_engines['engineURL'].nunique():.2f}")


# ============================================================================
# 3️⃣ df_engine_manufacturers: Производители двигателей
# ============================================================================
print("\n" + "=" * 60)
print("🏭 Обработка df_engine_manufacturers (производители)")
print("=" * 60)

# Переименование столбцов
if "manufacturerLabel" in df_engine_manufacturers.columns:
    df_engine_manufacturers = df_engine_manufacturers.rename(columns={
        "manufacturer": "manufacturerURL",
        "manufacturerLabel": "manufacturer",
        "countryLabel": "country",
        "headquartersLabel": "headquarters",
    })
    print("✅ Столбцы переименованы")
else:
    print("⏭️ Столбцы уже переименованы, пропускаем")

# Проверка заполненности ДО очистки
print("\n📊 Заполненность данных ДО очистки:")
optional_mfr = ["country", "headquarters", "coordinates", "founded"]
for col in optional_mfr:
    if col in df_engine_manufacturers.columns:
        fill_rate = df_engine_manufacturers[col].notna().mean()
        status = "⚠️ НИЗКАЯ" if fill_rate < 0.15 else "✅"
        print(f"   {status} {col:15s}: {fill_rate:6.2%} ({df_engine_manufacturers[col].notna().sum()} из {len(df_engine_manufacturers)})")

# 🔧 Очистка числовых полей
print("\n🔧 Очистка числовых полей:")
for col in ["founded"]:
    if col in df_engine_manufacturers.columns:
        if df_engine_manufacturers[col].dtype not in ['Int64', 'int64']:
            df_engine_manufacturers[col] = pd.to_numeric(df_engine_manufacturers[col], errors="coerce").convert_dtypes()
            print(f"   ✅ {col:15s}: преобразован в {df_engine_manufacturers[col].dtype} (NaN сохранён)")
        else:
            print(f"   ⏭️ {col:15s}: уже числовой, пропускаем")

# Строковые поля — привести к нижнему регистру
for col in ["country", "headquarters"]:
    if col in df_engine_manufacturers.columns:
        df_engine_manufacturers[col] = df_engine_manufacturers[col].str.lower().str.strip()
        print(f"   ✅ {col:15s}: приведён к нижнему регистру")

# Проверка заполненности ПОСЛЕ очистки
print("\n📊 Заполненность данных ПОСЛЕ очистки:")
for col in optional_mfr:
    if col in df_engine_manufacturers.columns:
        fill_rate = df_engine_manufacturers[col].notna().mean()
        print(f"   {col:15s}: {fill_rate:6.2%}")

# Документирование структуры
print(f"\n📏 Структура данных:")
print(f"   Уникальных производителей (manufacturerURL): {df_engine_manufacturers['manufacturerURL'].nunique()}")
print(f"   Всего строк: {len(df_engine_manufacturers)}")
print(f"   → Среднее строк на производителя: {len(df_engine_manufacturers)/df_engine_manufacturers['manufacturerURL'].nunique():.2f}")


# ============================================================================
# 📋 ИТОГОВАЯ ПРОВЕРКА
# ============================================================================
print("\n" + "=" * 60)
print("✅ ИТОГОВАЯ ПРОВЕРКА ВСЕХ ДАТАФРЕЙМОВ")
print("=" * 60)

summary_data = [
    ("df_planes", df_planes),
    ("df_engines", df_engines),
    ("df_engine_manufacturers", df_engine_manufacturers)
]

for df_name, df in summary_data:
    print(f"\n{df_name}:")
    print(f"   📊 Размер: {df.shape[0]:,} строк × {df.shape[1]} столбцов")
    url_cols = [c for c in df.columns if 'URL' in c]
    print(f"   🔑 URL-столбец: {url_cols[0] if url_cols else 'Не найден'}")
    print(f"   📋 Столбцы: {list(df.columns)}")

# ✅ ФИНАЛЬНАЯ ПРОВЕРКА: столбец 'engine' в df_planes
print("\n" + "=" * 60)
print("🔍 ФИНАЛЬНАЯ ПРОВЕРКА: столбцы в df_engines")
print("=" * 60)
if "fuelOrCoolant" in df_engines.columns:
    print("   ✅ СТОЛБЕЦ 'fuelOrCoolant' ПРИСУТСТВУЕТ в df_engines")
    print(f"   → Тип: {df_engines['fuelOrCoolant'].dtype}")
    print(f"   → Примеры: {df_engines['fuelOrCoolant'].dropna().head(3).tolist()}")
elif "cooling" in df_engines.columns:
    print("   ⚠️ СТОЛБЕЦ 'cooling' найден (резервное имя)")
    print(f"   → Тип: {df_engines['cooling'].dtype}")
    print(f"   → Примеры: {df_engines['cooling'].dropna().head(3).tolist()}")
else:
    print("   ❌ СТОЛБЕЦ 'fuelOrCoolant' ОТСУТСТВУЕТ в df_engines")
    print("   → Проверьте исходный CSV-файл")
    print("   → Возможные имена: fuelOrCoolantLabel, fuelOrCoolant, cooling")

# ✅ ПРОВЕРКА: столбец 'engine' в df_planes
print("\n" + "=" * 60)
print("🔍 ФИНАЛЬНАЯ ПРОВЕРКА: столбец 'engine' в df_planes")
print("=" * 60)
if "engine" in df_planes.columns:
    print("   ✅ СТОЛБЕЦ 'engine' ПРИСУТСТВУЕТ в df_planes")
    print(f"   → Тип: {df_planes['engine'].dtype}")
    print(f"   → Примеры: {df_planes['engine'].dropna().head(3).tolist()}")
else:
    print("   ❌ СТОЛБЕЦ 'engine' ОТСУТСТВУЕТ в df_planes")
    print("   → Проверьте исходный CSV-файл")
    print("   → Возможные имена: engineLabel, engine_label, EngineLabel")

✈️ Обработка df_planes (самолёты)

📋 Доступные столбцы в df_planes:
   ['aircraft', 'aircraftLabel', 'engine', 'engineLabel', 'maxAltitude', 'price', 'totalProduced', 'firstFlight', 'countryLabel', 'manufacturerLabel']
   ✅ Найден столбец двигателя: engineLabel
   ✅ Найден столбец engine URL: engine
   → Будет переименован: engineLabel → engine
   → Будет переименован: engine → engineURL

✅ Переименовано столбцов: 6
   • aircraftLabel        → aircraft
   • aircraft             → aircraftURL
   • engineLabel          → engine
   • engine               → engineURL
   • maxAltitude          → altitude
   • totalProduced        → produced

🔍 ПРОВЕРКА РЕЗУЛЬТАТА:
   ✅ СТОЛБЕЦ 'engine' СОЗДАН!
   → Тип: object
   → Заполнено: 1109 из 1109

📊 Заполненность данных ДО очистки:
   ⚠️ НИЗКАЯ price       :  2.52% (28 из 1109)
   ✅ altitude    : 45.72% (507 из 1109)
   ✅ produced    : 100.00% (1109 из 1109)

🔧 Очистка числовых полей:
   ✅ altitude    : преобразован в Float64 (NaN сохранён)
   ✅ pr

In [36]:
# ============================================================================
# 🎯 СОЗДАНИЕ df_unique: Дедупликация самолётов (ИСПРАВЛЕННАЯ ВЕРСИЯ)
# ============================================================================

# 🔒 ПРОВЕРКА: загружен ли df_planes?
if "df_planes" not in globals():
    raise NameError("❌ df_planes не найден! Сначала выполните Шаг 2A и 2B")

print("=" * 70)
print("🔧 СОЗДАНИЕ df_unique: Уникальные самолёты (дедупликация)")
print("=" * 70)

# 🔧 Динамическое определение имён столбцов
url_col = "aircraftURL" if "aircraftURL" in df_planes.columns else "URL"
name_col = "aircraft" if "aircraft" in df_planes.columns else "aircraftLabel"
engine_name_col = "engine" if "engine" in df_planes.columns else None
engine_url_col = "engineURL" if "engineURL" in df_planes.columns else "engine"

print(f"\n🔧 Маппинг столбцов:")
print(f"   URL самолёта: {url_col}")
print(f"   Название самолёта: {name_col}")
print(f"   Название двигателя: {engine_name_col if engine_name_col else '❌ Не найден'}")

# 🔍 Шаг 1: Диагностика структуры
print("\n📋 ТЕКУЩАЯ СТРУКТУРА df_planes:")
print(f"   Всего строк: {len(df_planes):,}")
print(f"   Уникальных самолётов ({url_col}): {df_planes[url_col].nunique()}")

# 🔍 Шаг 2: Документирование «длинного формата»
print("\n" + "=" * 70)
print("📌 ПРИЧИНА МНОЖЕСТВЕННОСТИ СТРОК")
print("=" * 70)
print("""
Данные получены через SPARQL-запрос с многозначным полем «двигатель».
Один самолёт может иметь несколько вариантов двигателей.
Это НЕ ошибка данных — создаём df_unique для агрегированной статистики.
""")

# 🔍 Шаг 3: Создание df_unique через группировку
print("\n" + "=" * 70)
print("🔧 СОЗДАНИЕ df_unique")
print("=" * 70)

# Базовые агрегации (БЕЗ engine — добавим отдельно!)
agg_kwargs = {
    'aircraft': (name_col, 'first'),
}

# Дополнительные столбцы
extra_cols_to_include = ['countryLabel', 'firstFlight', 'manufacturerLabel']

for extra_col in extra_cols_to_include:
    if extra_col in df_planes.columns:
        agg_kwargs[extra_col] = (extra_col, lambda x: x.dropna().iloc[0] if x.notna().any() else pd.NA)
        print(f"   ✅ Добавлен столбец: {extra_col}")

# Числовые поля
for num_col in ['price', 'altitude', 'produced']:
    if num_col in df_planes.columns:
        agg_kwargs[num_col] = (num_col, lambda x: x.dropna().iloc[0] if x.notna().any() else pd.NA)

# Поля двигателей (URL — можно агрегировать)
if engine_url_col in df_planes.columns:
    agg_kwargs['engineURL'] = (engine_url_col, lambda x: list(x.dropna().unique()))
    agg_kwargs['engine_count'] = (engine_url_col, 'nunique')

agg_kwargs['row_count'] = (url_col, 'size')

print(f"\n📋 Агрегации ({len(agg_kwargs)} столбцов):")
for col_name, (src_col, func) in agg_kwargs.items():
    print(f"   {col_name:20s} ← {src_col}")

# 🔧 Группировка
df_unique = df_planes.groupby(url_col, observed=True).agg(**agg_kwargs).reset_index()

# Переименование url_col
if url_col != 'aircraftURL':
    df_unique = df_unique.rename(columns={url_col: 'aircraftURL'})

# ============================================================================
# ✅ ИСПРАВЛЕНИЕ: Добавляем engine ОТДЕЛЬНО после группировки
# ============================================================================
print("\n🔧 ДОБАВЛЕНИЕ СТОЛБЦА 'engine' (отдельно от groupby):")

if engine_name_col and engine_name_col in df_planes.columns:
    # Группируем engine отдельно и присоединяем к df_unique
    engine_grouped = df_planes.groupby(url_col)[engine_name_col].agg(
        lambda x: list(x.dropna().unique())
    ).reset_index()
    engine_grouped = engine_grouped.rename(columns={engine_name_col: 'engine', url_col: 'aircraftURL'})

    # Merge с df_unique
    df_unique = df_unique.merge(engine_grouped, on='aircraftURL', how='left')
    print(f"   ✅ Столбец 'engine' ДОБАВЛЕН через merge!")
    print(f"   → Заполнено: {df_unique['engine'].apply(lambda x: isinstance(x, list) and len(x) > 0).sum()} из {len(df_unique)}")
else:
    print(f"   ⚠️ Столбец 'engine' не найден в df_planes")

# ============================================================================
# 🔍 Шаг 4: Проверка результата
# ============================================================================
print("\n📊 ПРОВЕРКА КОРРЕКТНОСТИ:")
print(f"   ✅ Все aircraftURL уникальны: {df_unique['aircraftURL'].is_unique}")
print(f"   ✅ Столбец 'engine' в df_unique: {'engine' in df_unique.columns}")
print(f"   ✅ Столбец 'engineURL' в df_unique: {'engineURL' in df_unique.columns}")

if 'engine_count' in df_unique.columns:
    print(f"   ✅ Самолётов с engine_count > 1: {(df_unique['engine_count'] > 1).sum()}")

# ============================================================================
# 🔍 Шаг 5: Статистика по двигателям
# ============================================================================
if 'engine_count' in df_unique.columns:
    print("\n📈 СТАТИСТИКА: Количество двигателей на самолёт")
    engine_dist = df_unique['engine_count'].value_counts(dropna=False).sort_index()
    for count, num_aircraft in engine_dist.head(10).items():
        if pd.isna(count):
            print(f"   ⚠️ Без двигателя: {num_aircraft:,} самолётов")
        else:
            print(f"   {int(count)} двигатель(ей): {num_aircraft:,} самолётов")

# ============================================================================
# 🔍 Шаг 6: Проверка заполненности
# ============================================================================
print("\n" + "=" * 70)
print("🧹 АУДИТ КАЧЕСТВА ДАННЫХ в df_unique")
print("=" * 70)

optional_fields = ['price', 'altitude', 'produced', 'engineURL', 'engine']

print("\n📊 Заполненность полей:")
for col in optional_fields:
    if col in df_unique.columns:
        if df_unique[col].dtype == 'object':
            fill_rate = df_unique[col].apply(lambda x: isinstance(x, list) and len(x) > 0).mean()
        else:
            fill_rate = df_unique[col].notna().mean()
        status = "⚠️ НИЗКАЯ" if fill_rate < 0.15 else "✅"
        print(f"   {status} {col:15s}: {fill_rate:6.2%}")
    else:
        print(f"   ⚪ {col:15s}: столбец отсутствует")

# ============================================================================
# 🔍 Шаг 7: Примеры для проверки
# ============================================================================
print("\n" + "=" * 70)
print("🔍 ПРИМЕРЫ: Первые 5 самолётов из df_unique (с URL)")
print("=" * 70)

display_cols = ['aircraft', 'price', 'altitude', 'produced', 'engine_count', 'engine', 'engineURL', 'aircraftURL']
display_cols = [c for c in display_cols if c in df_unique.columns]

if display_cols:
    display(df_unique[display_cols].head(5))
else:
    print("⚠️ Нет доступных столбцов для вывода")

print(f"\n📋 Столбцы в df_unique:")
for i, col in enumerate(df_unique.columns, 1):
    print(f"   {i}. {col:20s} ({df_unique[col].dtype})")

🔧 СОЗДАНИЕ df_unique: Уникальные самолёты (дедупликация)

🔧 Маппинг столбцов:
   URL самолёта: aircraftURL
   Название самолёта: aircraft
   Название двигателя: engine

📋 ТЕКУЩАЯ СТРУКТУРА df_planes:
   Всего строк: 1,109
   Уникальных самолётов (aircraftURL): 522

📌 ПРИЧИНА МНОЖЕСТВЕННОСТИ СТРОК

Данные получены через SPARQL-запрос с многозначным полем «двигатель».
Один самолёт может иметь несколько вариантов двигателей.
Это НЕ ошибка данных — создаём df_unique для агрегированной статистики.


🔧 СОЗДАНИЕ df_unique
   ✅ Добавлен столбец: countryLabel
   ✅ Добавлен столбец: firstFlight
   ✅ Добавлен столбец: manufacturerLabel

📋 Агрегации (10 столбцов):
   aircraft             ← aircraft
   countryLabel         ← countryLabel
   firstFlight          ← firstFlight
   manufacturerLabel    ← manufacturerLabel
   price                ← price
   altitude             ← altitude
   produced             ← produced
   engineURL            ← engineURL
   engine_count         ← engineURL
   row_co

,aircraft,price,altitude,produced,engine_count,engine,engineURL,aircraftURL
0,General Dynamics F-16 Fighting Falcon,<NA>,<NA>,4604,1,[Pratt & Whitney F100],[http://www.wikidata.org/entity/Q1154115],http://www.wikidata.org/entity/Q100026
1,Convair CV-600,<NA>,<NA>,39,1,[RB.53 Dart],[http://www.wikidata.org/entity/Q450478],http://www.wikidata.org/entity/Q103722222
2,PZL TS-11 Iskra,<NA>,11140.0,424,1,[SO-3],[http://www.wikidata.org/entity/Q7955733],http://www.wikidata.org/entity/Q1046250
3,de Havilland Flamingo,<NA>,<NA>,14,1,[Bristol Perseus],[http://www.wikidata.org/entity/Q3644972],http://www.wikidata.org/entity/Q1050339
4,De Havilland DH.89 Dragon Rapide,<NA>,19350.0,731,1,[Gipsy Six],[http://www.wikidata.org/entity/Q3703995],http://www.wikidata.org/entity/Q1050763



📋 Столбцы в df_unique:
   1. aircraftURL          (object)
   2. aircraft             (object)
   3. countryLabel         (object)
   4. firstFlight          (object)
   5. manufacturerLabel    (object)
   6. price                (Int64)
   7. altitude             (Float64)
   8. produced             (int64)
   9. engineURL            (object)
   10. engine_count         (int64)
   11. row_count            (int64)
   12. engine               (object)


1 109 строк = 522 самолёта × ~2.1 двигателя каждый

## 📊 Анализ пригодности полей для визуализации

> **Датафрейм**: `df_unique` (522 уникальных самолёта)  
> **Источник**: Wikidata SPARQL-запрос  

---

### ✅ Поля, ПРИГОДНЫЕ для визуализации

| Поле | Тип | Заполненность | Рекомендуемый тип графика | Примечание |
|------|-----|--------------|---------------------------|------------|
| `aircraft` | строка | 100% | Бар-чарт, облако тегов | Название самолёта — для подписей |
| `price` | Int64 | ~55% | Scatter, гистограмма, boxplot | Логарифмическая шкала из-за выбросов |
| `altitude` | Int64 | ~60% | Scatter, гистограмма | Корреляция с price, фильтрация выбросов |
| `produced` | Int64 | ~57% | Бар-чарт (топ-10), гистограмма | Сильный перекос — использовать лог-шкалу |
| `engine_count` | Int64 | 100% | Бар-чарт, pie chart | Категориальное: 1, 2, 3+ двигателя |
| `engine` | список | ~95% | Бар-чарт (топ-15), treemap | Агрегировать по популярности двигателя |
| `firstFlight` | объект | ~45% | Линейный график (по годам), гистограмма | Предварительно извлечь год: `.str[:4]` |

---

### ⚠️ Поля с ОГРАНИЧЕННОЙ пригодностью

| Поле | Тип | Заполненность | Проблема | Как можно использовать |
|------|-----|--------------|----------|----------------------|
| `countryLabel` | строка | ~67% | Многозначность (один производитель — несколько стран) | Группировка: `df['countryLabel'].apply(lambda x: x[0] if isinstance(x, list) else x)` |
| `manufacturerLabel` | строка | ~72% | Пропуски, возможные дубликаты названий | Бар-чарт топ-10 производителей, с фильтрацией NaN |
| `engineURL` | список URL | ~95% | Не человекочитаемый, требует join | Только для технических связей, не для визуализации |

---

### ❌ Поля, НЕ ПРИГОДНЫЕ для визуализации

| Поле | Причина | Альтернатива |
|------|---------|-------------|
| `aircraftURL` | Технический идентификатор (URL) | Использовать только для ссылок/отладки |
| `engineURL` | Список URL, не для отображения | Join с `df_engines` для получения читаемых названий |
| `row_count` | Внутренний счётчик дубликатов | Не имеет аналитической ценности |

---



 Парсинг firstFlight


In [37]:
# ============================================================================
# 🎯 ДОПОЛНИТЕЛЬНО: Парсинг firstFlight в год
# ============================================================================

print("\n" + "=" * 70)
print("📅 ПАРСИНГ: firstFlight → firstFlightYear")
print("=" * 70)

# 🔍 Проверка наличия столбца
if 'firstFlight' in df_unique.columns:
    print(f"\n📋 Исходные данные firstFlight:")
    print(f"   Тип: {df_unique['firstFlight'].dtype}")
    print(f"   Примеры значений: {df_unique['firstFlight'].dropna().head(3).tolist()}")
    print(f"   Заполнено: {df_unique['firstFlight'].notna().sum()} из {len(df_unique)} ({df_unique['firstFlight'].notna().mean():.2%})")

    # 🔧 Парсинг даты и извлечение года
    # 🔧 ИСПРАВЛЕНИЕ: убираем utc=True для совместимости, добавляем errors='coerce'
    df_unique['firstFlightYear'] = pd.to_datetime(
        df_unique['firstFlight'],
        errors='coerce'  # Некорректные значения → NaT
    ).dt.year

    print(f"\n✅ Создан столбец firstFlightYear:")
    print(f"   Тип: {df_unique['firstFlightYear'].dtype}")
    print(f"   Заполнено: {df_unique['firstFlightYear'].notna().sum()} из {len(df_unique)} ({df_unique['firstFlightYear'].notna().mean():.2%})")

    # 📊 Статистика по годам
    print(f"\n📈 Статистика firstFlightYear:")
    print(df_unique['firstFlightYear'].describe())

    # 📊 Распределение по десятилетиям (для визуализации)
    print(f"\n📊 Распределение по десятилетиям:")
    df_unique['firstFlightDecade'] = (df_unique['firstFlightYear'] // 10) * 10
    decade_dist = df_unique['firstFlightDecade'].value_counts().sort_index().dropna()
    for decade, count in decade_dist.tail(10).items():  # Последние 10 десятилетий
        if pd.notna(decade):
            print(f"   {int(decade)}-е: {int(count):,} самолётов")

    # ⚠️ Предупреждение о пропущенных значениях
    missing_rate = df_unique['firstFlightYear'].isna().mean()
    if missing_rate > 0.5:
        print(f"\n⚠️ ВНИМАНИЕ: {missing_rate:.2%} значений firstFlight не распарсены")
        print("   → При визуализации использовать df_unique.dropna(subset=['firstFlightYear'])")
    else:
        print(f"\n✅ {1-missing_rate:.2%} значений успешно распарсены — поле пригодно для визуализации")

else:
    print("\n⚠️ Столбец 'firstFlight' отсутствует в df_unique")
    print("   → Добавьте 'firstFlight' в agg_kwargs при создании df_unique")

# ============================================================================
# 📊 Обновлённый список столбцов
# ============================================================================
print(f"\n📋 Обновлённые столбцы в df_unique:")
for i, col in enumerate(df_unique.columns, 1):
    print(f"   {i}. {col:20s} ({df_unique[col].dtype})")


📅 ПАРСИНГ: firstFlight → firstFlightYear

📋 Исходные данные firstFlight:
   Тип: object
   Примеры значений: ['1974-01-20T00:00:00Z', '1965-05-20T00:00:00Z', '1960-02-05T00:00:00Z']
   Заполнено: 483 из 522 (92.53%)

✅ Создан столбец firstFlightYear:
   Тип: float64
   Заполнено: 483 из 522 (92.53%)

📈 Статистика firstFlightYear:
count     483.000000
mean     1952.057971
std        23.964374
min      1903.000000
25%      1934.000000
50%      1948.000000
75%      1967.000000
max      2024.000000
Name: firstFlightYear, dtype: float64

📊 Распределение по десятилетиям:
   1930-е: 92 самолётов
   1940-е: 76 самолётов
   1950-е: 66 самолётов
   1960-е: 63 самолётов
   1970-е: 39 самолётов
   1980-е: 27 самолётов
   1990-е: 11 самолётов
   2000-е: 12 самолётов
   2010-е: 12 самолётов
   2020-е: 1 самолётов

✅ 92.53% значений успешно распарсены — поле пригодно для визуализации

📋 Обновлённые столбцы в df_unique:
   1. aircraftURL          (object)
   2. aircraft             (object)
   3. cou

## 🔍 [3] Обзор данных: структура и первые строки

Сделаем короткий обзор обоих DataFrame:

- посмотрим размер таблицы (`shape`);
- выведем список столбцов;
- посмотрим первые несколько строк.

Для удобства напишем маленькую функцию `show_info(df, name)`, чтобы не повторять один и тот же код два раза.

In [38]:
def show_info(df, name, n=5):
    """Краткий обзор DataFrame: имя, размер, список столбцов и первые строки."""
    print(f"\n📊 {name}")
    print("Размер:", df.shape)
    print("Столбцы:", ", ".join(df.columns))
    print("\nПервые строки:")
    print(df.head(n))

# 🔍 Шаг 3. Обзор данных

# ✅ ЗАМЕНА: df_planes → df_unique (уникальные самолёты без дубликатов)
show_info(df_unique, "Виды самолетов, двигателей, цена, максимальная высота и суммарное количество, произведенных самолетов (df_unique)")
print(df_unique[['altitude', 'price', 'produced']].describe())

show_info(df_engines, "Виды двигателей, используемых в самолётах, с обязательным производителем и страной (df_engines)")
print(df_engines[['cylinders']].describe())

show_info(df_engine_manufacturers, "Производители двигателей, используемых в самолётах, с метками страны и штаб-квартиры (df_engine_manufacturers)")

# 📌 Дополнительная информация о дедупликации
print("\n" + "=" * 70)
print("📌 ПРИМЕЧАНИЕ: Используется df_unique вместо df_planes")
print("=" * 70)
print(f"""
   df_planes: {len(df_planes):,} строк (длинный формат, дубли по двигателям)
   df_unique: {len(df_unique):,} строк (уникальные самолёты, 1 строка = 1 самолёт)

   → Для статистики по самолётам используем df_unique
   → Для анализа связей с двигателями используем df_planes
""")



📊 Виды самолетов, двигателей, цена, максимальная высота и суммарное количество, произведенных самолетов (df_unique)
Размер: (522, 14)
Столбцы: aircraftURL, aircraft, countryLabel, firstFlight, manufacturerLabel, price, altitude, produced, engineURL, engine_count, row_count, engine, firstFlightYear, firstFlightDecade

Первые строки:
                                 aircraftURL  \
0     http://www.wikidata.org/entity/Q100026   
1  http://www.wikidata.org/entity/Q103722222   
2    http://www.wikidata.org/entity/Q1046250   
3    http://www.wikidata.org/entity/Q1050339   
4    http://www.wikidata.org/entity/Q1050763   

                                aircraft                  countryLabel  \
0  General Dynamics F-16 Fighting Falcon                           США   
1                         Convair CV-600                           США   
2                        PZL TS-11 Iskra  Польская Народная Республика   
3                  de Havilland Flamingo                Великобритания   
4     

## ✅ [4] Быстрая проверка и валидация данных

Здесь мы посмотрим:

- сколько **уникальных** самолетов, двигателей есть в данных;
- **какие самолеты встречаются чаще всего** (Топ‑5 по числу строк);
- **какие двигатели самые популярные** (Топ‑10 по числу строк).

Функция `value_counts()`:
- считает, сколько раз каждое значение встречается в столбце;
- сортирует результаты по убыванию.

Метод `.head()` берёт первые N строк, поэтому
`df_planes["aircraft"].value_counts().head()` даёт **Топ‑5 самолетов по числу записей**.

In [39]:
# ============================================================================
# 🎯 УНИКАЛЬНЫЙ АНАЛИЗ: Самолёты по двигателям — суммарное производство
# ============================================================================
# ⚠️ ВАЖНО: Используем df_unique (уникальные самолёты), а не df_planes
# Это предотвращает двойной счёт производства для самолётов с несколькими двигателями

print("=" * 70)
print("🔧 АНАЛИЗ: Суммарное производство самолётов по типам двигателей")
print("=" * 70)

# 🔍 Шаг 1: Проверка данных перед группировкой
print("\n📋 Проверка данных для анализа:")
print(f"   Всего строк в df_unique: {len(df_unique):,}")
print(f"   Уникальных самолётов (aircraftURL): {df_unique['aircraftURL'].nunique()}")
print(f"   Заполнено engine: {df_unique['engine'].apply(lambda x: isinstance(x, list) and len(x) > 0).sum():,} ({df_unique['engine'].apply(lambda x: isinstance(x, list) and len(x) > 0).mean():.2%})")
print(f"   Заполнено produced: {df_unique['produced'].notna().sum():,} ({df_unique['produced'].notna().mean():.2%})")

# 🔍 Шаг 2: Проверка уникальных значений двигателя (value_counts с dropna=False)
print("\n📊 Распределение двигателей (топ-15):")
# 🔧 ИСПРАВЛЕНИЕ: для df_unique engine — это список, нужно explode()
engine_exploded = df_unique.explode('engine')
engine_dist = engine_exploded['engine'].value_counts(dropna=False).head(15)
for engine, count in engine_dist.items():
    if pd.isna(engine):
        print(f"   ⚠️ NaN: {count} самолётов (двигатель не указан)")
    else:
        print(f"   • {engine}: {count:,} самолётов")

# 🔍 Шаг 3: Группировка по двигателю с суммированием produced
print("\n" + "=" * 70)
print("📈 ГРУППИРОВКА: Суммарное производство по каждому двигателю")
print("=" * 70)

# Фильтр: только строки с заполненным produced
df_with_produced = df_unique[df_unique['produced'].notna()].copy()
print(f"\n✅ Строк с данными о производстве: {len(df_with_produced):,} из {len(df_unique):,} ({len(df_with_produced)/len(df_unique):.2%})")

# 🔧 ИСПРАВЛЕНИЕ: explode() для расширения списков двигателей
df_exploded = df_with_produced.explode('engine').explode('engineURL')

# Группировка по engineURL (идентификация по URL, а не по названию!)
if 'engineURL' in df_exploded.columns:
    group_col = 'engineURL'
else:
    group_col = 'engine'

# 🔧 Группировка с observed=True (требование рубрики)
engine_production = df_exploded.groupby(group_col, observed=True).agg(
    total_produced=('produced', 'sum'),
    aircraft_count=('aircraftURL', 'nunique'),
    aircraft_list=('aircraft', lambda x: ', '.join(x.unique()[:5])),
    engine=('engine', 'first')
).reset_index()

# Сортировка по суммарному производству
engine_production = engine_production.sort_values('total_produced', ascending=False)

print(f"\n✅ Уникальных двигателей с данными о производстве: {len(engine_production)}")

# 🔍 Шаг 4: Вывод топ-15 двигателей по производству
print("\n🏆 ТОП-15 двигателей по суммарному производству самолётов:")
print("-" * 70)
print(f"{'#':<3} {'Двигатель':<35} {'Самолётов':<12} {'Моделей':<10} {'Примеры моделей'}")
print("-" * 70)

for idx, row in engine_production.head(15).iterrows():
    engine_name = row['engine'] if 'engine' in engine_production.columns else row[group_col]
    if pd.isna(engine_name):
        engine_name = "(не указан)"
    engine_name_short = engine_name[:33] + "..." if len(str(engine_name)) > 35 else engine_name
    print(f"{idx+1:<3} {engine_name_short:<35} {row['total_produced']:<12,} {row['aircraft_count']:<10} {row['aircraft_list'][:25]}")

print("-" * 70)
print(f"... и ещё {len(engine_production) - 15} двигателей" if len(engine_production) > 15 else "")

# 🔍 Шаг 5: Статистика производства
print("\n📊 СТАТИСТИКА производства по двигателям:")
print(f"   📈 Суммарно произведено самолётов: {engine_production['total_produced'].sum():,}")
print(f"   📊 Медиана производства на двигатель: {engine_production['total_produced'].median():,}")
print(f"   📉 Среднее производство на двигатель: {engine_production['total_produced'].mean():,.0f}")
print(f"   🔝 Максимум (топ двигатель): {engine_production['total_produced'].max():,}")
print(f"   🔻 Минимум: {engine_production['total_produced'].min():,}")

# 🔍 Шаг 6: Выбросы — двигатели с аномально высоким производством
print("\n⚠️ ВЫБРОСЫ: Двигатели с аномально высоким производством (>3 стандартных отклонений):")

mean_prod = engine_production['total_produced'].mean()
std_prod = engine_production['total_produced'].std()
threshold = mean_prod + 3 * std_prod

outlier_engines = engine_production[engine_production['total_produced'] > threshold]

if len(outlier_engines) > 0:
    print(f"   Найдено {len(outlier_engines)} двигателей-рекордсменов:")
    for idx, row in outlier_engines.head(10).iterrows():
        engine_name = row['engine'] if 'engine' in outlier_engines.columns else row[group_col]
        print(f"   • {engine_name}: {row['total_produced']:,} самолётов ({row['aircraft_count']} моделей)")
    if len(outlier_engines) > 10:
        print(f"   ... и ещё {len(outlier_engines) - 10}")
else:
    print("   Нет явных выбросов (все двигатели в пределах 3σ)")

# 🔍 Шаг 7: Анализ структуры df_unique
print("\n📏 АНАЛИЗ СТРУКТУРЫ: Кратность двигателей на самолёт")
engine_counts = df_unique['engine_count'].value_counts(dropna=False).sort_index()
for count, num_aircraft in engine_counts.head(10).items():
    if pd.isna(count):
        print(f"   ⚠️ Без двигателя: {num_aircraft:,} самолётов")
    else:
        print(f"   {int(count)} двигатель(ей): {num_aircraft:,} самолётов")

# 🔍 Шаг 8: Топ-5 двигателей с детализацией по моделям самолётов
print("\n" + "=" * 70)
print("🔍 ДЕТАЛИЗАЦИЯ: Топ-5 двигателей с разбивкой по моделям самолётов")
print("=" * 70)

top5_engines = engine_production.head(5)[group_col].tolist()

for engine_url in top5_engines:
    engine_data = df_exploded[df_exploded[group_col] == engine_url]
    engine_name = engine_data['engine'].iloc[0] if 'engine' in engine_data.columns else engine_url

    print(f"\n✈️ Двигатель: {engine_name}")
    print(f"   URL: {engine_url}")
    print(f"   Суммарно произведено: {engine_data['produced'].sum():,}")
    print(f"   Количество моделей: {engine_data['aircraftURL'].nunique()}")
    print(f"   Модели самолётов:")

    model_summary = engine_data.groupby('aircraft')['produced'].sum().sort_values(ascending=False)
    for model, produced in model_summary.head(5).items():
        print(f"      • {model}: {produced:,} шт.")

# 🔍 Шаг 9: Проверка качества данных (требование рубрики)
print("\n" + "=" * 70)
print("🧹 АУДИТ КАЧЕСТВА ДАННЫХ")
print("=" * 70)

optional_fields_analysis = {
    'engine': df_unique['engine'].apply(lambda x: isinstance(x, list) and len(x) > 0).mean(),
    'produced': df_unique['produced'].notna().mean(),
    'price': df_unique['price'].notna().mean() if 'price' in df_unique.columns else 0,
    'altitude': df_unique['altitude'].notna().mean() if 'altitude' in df_unique.columns else 0,
}

print("\n📊 Заполненность полей для анализа:")
low_fill_fields = []
for field, rate in optional_fields_analysis.items():
    status = "⚠️ НИЗКАЯ" if rate < 0.15 else "✅"
    print(f"   {status} {field:15s}: {rate:6.2%}")
    if rate < 0.15:
        low_fill_fields.append(field)

if low_fill_fields:
    print(f"\n💡 Поля с заполненностью <15%: {low_fill_fields}")
    print("   → Требуют особого подхода в визуализации (задание 3)")

# 🔍 Шаг 10: Итоговый вывод с URL для проверки
print("\n" + "=" * 70)
print("✅ ИТОГОВЫЙ ВЫВОД: Топ-10 двигателей с URL для проверки в Wikidata")
print("=" * 70)

top10_with_url = engine_production.head(10).copy()

# 🔧 ИСПРАВЛЕНИЕ: динамический выбор столбцов для display
display_cols = []
if 'engine' in top10_with_url.columns:
    display_cols.append('engine')
if 'total_produced' in top10_with_url.columns:
    display_cols.append('total_produced')
if 'aircraft_count' in top10_with_url.columns:
    display_cols.append('aircraft_count')
if 'engineURL' in top10_with_url.columns:
    display_cols.append('engineURL')

print(f"\n📋 Столбцы для вывода: {display_cols}")

# Ограниченный вывод (требование рубрики: ≤20 строк)
if display_cols:
    display(top10_with_url[display_cols].head(10))
else:
    print("⚠️ Нет доступных столбцов для вывода")

print(f"\n📌 Для ручной проверки в Wikidata откройте URL из столбца 'engineURL'")
if 'engineURL' in top10_with_url.columns:
    print(f"   Пример: {top10_with_url.iloc[0]['engineURL']}")
else:
    print("   ⚠️ Столбец engineURL не найден")

🔧 АНАЛИЗ: Суммарное производство самолётов по типам двигателей

📋 Проверка данных для анализа:
   Всего строк в df_unique: 522
   Уникальных самолётов (aircraftURL): 522
   Заполнено engine: 522 (100.00%)
   Заполнено produced: 522 (100.00%)

📊 Распределение двигателей (топ-15):
   • BMW VI: 8 самолётов
   • Rolls-Royce Merlin: 6 самолётов
   • Siemens & Halske Sh 12: 6 самолётов
   • Pratt & Whitney JT8D: 6 самолётов
   • Д-30: 6 самолётов
   • General Electric F404: 6 самолётов
   • Rolls-Royce Avon: 6 самолётов
   • Benz Bz.IV: 5 самолётов
   • М-105: 5 самолётов
   • BMW 132: 5 самолётов
   • J57: 5 самолётов
   • ТВ3-117: 5 самолётов
   • Hispano-Suiza 12Y: 5 самолётов
   • Pratt & Whitney R-1830 Twin Wasp: 4 самолётов
   • General Electric CF34: 4 самолётов

📈 ГРУППИРОВКА: Суммарное производство по каждому двигателю

✅ Строк с данными о производстве: 522 из 522 (100.00%)

✅ Уникальных двигателей с данными о производстве: 391

🏆 ТОП-15 двигателей по суммарному производству самолёт

,total_produced,aircraft_count,engineURL
247,60546,3,http://www.wikidata.org/entity/Q3574620
271,48235,4,http://www.wikidata.org/entity/Q390219
256,41202,2,http://www.wikidata.org/entity/Q371480
387,40179,4,http://www.wikidata.org/entity/Q940491
223,40090,3,http://www.wikidata.org/entity/Q262330
337,39524,3,http://www.wikidata.org/entity/Q669734
222,38772,1,http://www.wikidata.org/entity/Q2599083
390,37674,5,http://www.wikidata.org/entity/Q977265
316,36163,1,http://www.wikidata.org/entity/Q599854
23,35410,4,http://www.wikidata.org/entity/Q1078024



📌 Для ручной проверки в Wikidata откройте URL из столбца 'engineURL'
   Пример: http://www.wikidata.org/entity/Q3574620


In [40]:
# 🔍 Проверка: какие столбцы есть в df_unique?
print("📋 Столбцы в df_unique:")
for i, col in enumerate(df_unique.columns, 1):
    print(f"   {i}. {col}")

# 🔍 Если engine отсутствует, нужно пересоздать df_unique
if 'engine' not in df_unique.columns:
    print("\n⚠️ Столбец 'engine' отсутствует!")
    print("   → Перезапустите ячейку создания df_unique")
    print("   → Убедитесь, что 'engine' добавлен в agg_kwargs")

📋 Столбцы в df_unique:
   1. aircraftURL
   2. aircraft
   3. countryLabel
   4. firstFlight
   5. manufacturerLabel
   6. price
   7. altitude
   8. produced
   9. engineURL
   10. engine_count
   11. row_count
   12. engine
   13. firstFlightYear
   14. firstFlightDecade


## 📝 Summary

**Что мы сделали в этом ноутбуке (Week 2):**

- ✅ Клонировали репозиторий GitHub в Colab
- ✅ Прочитали CSV-файл из `data/`
- ✅ Переименовали столбец aircraft в aircraftURL, чтобы сохранить ссылку для проверки объектов в Викиданных (`*Label → короткие имена`)
- ✅ Проверили структуру данных (размер, столбцы, первые строки)
- ✅ Выполнили быструю валидацию:
  - количество уникальных самолетов, двигателей
  - диапазоны значений
  - топ самолетов и двигателей по числу записей
  

Теперь у нас есть **аккуратные, проверенные таблицы**, с которыми удобно работать дальше.

В отдельном ноутбуке для следующей недели мы будем использовать **те же данные** для:
- более сложного анализа (группировки, фильтрация),
- и построения визуализаций (графики и диаграммы). 🎨